Index: lib/keyboard.py
===================================================================
--- lib/keyboard.py	(revision 596)
+++ lib/keyboard.py	(working copy)
@@ -186,3 +186,15 @@
 
         if os.access(argv[0], os.X_OK) == 1:
             call (argv)
+
+    def dracutSetupString(self):
+        args = ""
+        keyboardtype = self._var("KEYBOARDTYPE")
+        keytable = self._var("KEYTABLE")
+
+        if keyboardtype:
+             args += " KEYBOARDTYPE=%s" % keyboardtype
+        if keytable:
+             args += " KEYTABLE=%s" % keytable
+
+        return args
