Index: src/keyboard_backend.py
===================================================================
--- src/keyboard_backend.py	(revision 559)
+++ src/keyboard_backend.py	(working copy)
@@ -1,9 +1,11 @@
 #
-# keyboard_backend.py - backend code for mouse configuration
+# keyboard_backend.py - backend code for keyboard configuration
 #
 # Copyright (C) 2002, 2003 Red Hat, Inc.
 # Brent Fox <bfox@redhat.com>
 #
+# Copyright (C) 2008 Lubomir Kundrak <lkundrak@redhat.com>
+#
 # This program is free software; you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
 # the Free Software Foundation; either version 2 of the License, or
@@ -31,7 +33,18 @@
         if os.access("/etc/X11/XF86Config", os.W_OK) or os.access("/etc/X11/xorg.conf", os.W_OK):        
             import xf86config
             (xconfig, xconfigpath) = xf86config.readConfigFile()
-            keyboard = xf86config.getCoreKeyboard(xconfig)
+            try:
+                keyboard = xf86config.getCoreKeyboard(xconfig)
+            except:
+                xconfig.comment = '# This configuration file was broken by system-config-keyboard'
+                keyboard = xf86config.XF86ConfInput ();
+                keyboard.comment = "# Keyboard added by system-config-keyboard"
+                keyboard.identifier = "Keyboard0"
+                keyboard.driver = "kbd"
+                keyboard.options.insert (xf86config.XF86Option("XkbModel", "pc101"))
+                keyboard.options.insert (xf86config.XF86Option("XkbLayout", "us"))
+                xconfig.input.insert (keyboard)
+                xconfig.layout[0].inputs.insert (xf86config.XF86ConfInputref ("Keyboard0", "CoreKeyboard"));
             
             found = 0
             for o in keyboard.options:
